tutorials/misc/migrating-to-0.4.0.md

```diff
@@ -0,0 +1,139 @@
+`allennlp` 0.4.0 introduces several breaking changes.
+A couple of them you are likely to run into,
+most of them you are not.
+
+# CHANGES THAT YOU ARE LIKELY TO RUN INTO
+
+## `allennlp.run evaluate`
+
+The `--archive-file` parameter for `evaluate` was changed to a positional parameter,
+as it is in our other commands. This means that your previous
+
+```bash
+python -m allennlp.run evaluate --archive-file model.tar.gz --evaluation-data-file data.txt
+```
+
+will need to be changed to
+
+```bash
+python -m allennlp.run evaluate model.tar.gz --evaluation-data-file data.txt
+```
+
+## `DatasetReader`
+
+If you have written your own `DatasetReader` subclass, it will need a few small changes.
+
+Previously every dataset reader was *eager*;
+that is, it returned a `Dataset` object that wrapped a `list` of `Instance`s
+that necessarily stayed in memory throughout the training process.
+
+In 0.4.0 we enabled *lazy* datasets, which required some minor changes in
+how `DatasetReader`s work.
+
+Previously you would have overridden the `read` function with something like
+
+```python
+def read(self, file_path: str) -> Dataset:
+    instances: List[Instance] = []
+    # open some data_file
+    for line in data_file:
+       instance = ...
+       instances.append(instance)
+    return Dataset(instances)
+```
+
+In 0.4.0 the `read()` function is already implemented in the base `DatasetReader`
+class in a way that allows every dataset reader to generate instances either
+_lazily_ (that is, as needed, and anew each iteration) or
+_eagerly_ (that is, generated once and stored in a list)
+in a way that's transparent to the user.
+
+So in 0.4.0 your `DatasetReader` subclass needs to instead
+override the private `_read` function
+and `yield` instances one at time.
+
+This should only be like a 2-line change:
+
+```python
+# new name _read and return type: Iterable[Instance]
+def _read(self, file_path: str) -> Iterable[Instance]:
+    # open some data_file
+    for line in data_file:
+       instance = ...
+       # yield instances rather than collecting them in a list
+       yield instance
+```
+
+In addition, the base `DatasetReader` constructor now takes a `lazy: bool` parameter,
+which means that your subclass constructor should also take that parameter
+(unless you don't want to allow laziness, but why wouldn't you?)
+and explicitly pass it the superclass constructor:
+
+```python
+class MyDatasetReader(DatasetReader)
+    def __init__(self,
+                 # other arguments
+                 lazy: bool = False):
+        super().__init__(lazy)
+        # whatever other initialization you need
+```
+
+(For the reasoning behind this change, see the [Laziness tutorial](https://github.com/allenai/allennlp/blob/master/tutorials/getting_started/laziness.md).)
+
+# CHANGES YOU ARE MUCH LESS LIKELY TO RUN INTO
+
+If you only ever use our the command line tools (`python -m allennlp.run ...`) to train / evaluate models,
```

Comment:
You could say something in here about writing model / dataset reader code ("use our command line tools" doesn't make it obvious that you also shouldn't have to worry about anything else if you only wrote `Models` or `DatasetReaders`).  Also, you have "use our the command line tools".